src/Adapter/Order/CommandHandler/CancelOrderProductHandler.php

```diff
@@ -114,29 +110,17 @@ public function handle(CancelOrderProductCommand $command)
         if (!empty($orderDetails['productsOrderDetails'])) {
             foreach ($orderDetails['productsOrderDetails'] as $orderDetail) {
                 $qty_cancel_product = $orderDetails['productCancelQuantity'][$orderDetail->id_order_detail];
-                $this->reinjectQuantity($orderDetail, $qty_cancel_product);
-
-                $this->orderProductRemover->deleteProductFromOrder($order, $orderDetail, $qty_cancel_product);
-
-                // Update weight SUM
-                $order_carrier = new OrderCarrier((int) $order->getIdOrderCarrier());
-                if (Validate::isLoadedObject($order_carrier)) {
-                    $order_carrier->weight = (float) $order->getTotalWeight();
-                    if ($order_carrier->update()) {
-                        $order->weight = sprintf('%.3f ' . Configuration::get('PS_WEIGHT_UNIT'), $order_carrier->weight);
-                    }
-                }
-
-                if (Configuration::get('PS_ADVANCED_STOCK_MANAGEMENT') && StockAvailable::dependsOnStock($orderDetail->product_id)) {
-                    StockAvailable::synchronize($orderDetail->product_id);
-                }
+                $newQuantity = (int) $orderDetail->product_quantity - (int) $qty_cancel_product;
+                $this->orderProductQuantityUpdater->update($order, $orderDetail, $newQuantity, null);
                 Hook::exec('actionProductCancel', ['order' => $order, 'id_order_detail' => (int) $orderDetail->id_order_detail], null, false, true, false, $order->id_shop);
             }
         }
         if (!empty($orderDetails['customizedProductsOrderDetail'])) {
             foreach ($orderDetails['customizedProductsOrderDetail'] as $orderDetail) {
                 $qtyCancelProduct = abs($orderDetails['customizedCancelQuantity'][$orderDetail->id_customization]);
-                $this->orderProductRemover->deleteProductFromOrder($order, $orderDetail, $qtyCancelProduct);
+                $newQuantity = (int) $orderDetail->product_quantity - (int) $qtyCancelProduct;
+
+                $this->orderProductQuantityUpdater->update($order, $orderDetail, $newQuantity, null);
```

Comment:
Same here, you should provide the invoice if present (it can be generated manually even if the order has not been paid)

Reply:
same as above